src/poke_env/player/gymnasium_api.py

```diff
@@ -270,149 +305,161 @@ def action_space_size(self) -> int:
             else:
                 raise RuntimeError(
                     "Environment and agent aren't synchronized. Try to restart"
                 )
-        while self.current_battle == self.agent.current_battle:
+        while self.current_battle1 == self.agent1.current_battle:
             time.sleep(0.01)
-        self.current_battle = self.agent.current_battle
-        self.current_battle.logger = None
-        self.last_battle = self.current_battle
-        return self._observations.get(), self.get_additional_info()
-
-    def get_additional_info(self) -> Dict[str, Any]:
+        observations = {
+            self.agents[0]: self._observations1.get(),
+            self.agents[1]: self._observations2.get(),
+        }
+        self.current_battle1 = self.agent1.current_battle
+        self.current_battle1.logger = None
+        self.current_battle2 = self.agent2.current_battle
+        self.current_battle2.logger = None
+        self.last_battle1 = self.current_battle1
+        self.last_battle2 = self.current_battle2
+        return observations, self.get_additional_info()
+
+    def get_additional_info(self) -> Dict[str, Dict[str, Any]]:
         """
         Returns additional info for the reset method.
         Override only if you really need it.
 
         :return: Additional information as a Dict
         :rtype: Dict
         """
-        return {}
-
-    def step(
-        self, action: ActType
-    ) -> Tuple[ObsType, float, bool, bool, Dict[str, Any]]:
-        """
-        Execute the specified action in the environment.
-
-        :param ActType action: The action to be executed.
-        :return: A tuple containing the new observation, reward, termination flag, truncation flag, and info dictionary.
-        :rtype: Tuple[ObsType, float, bool, bool, Dict[str, Any]]
-        """
-        if not self.current_battle:
-            obs, info = self.reset()
-            return obs, 0.0, False, False, info
-        if self.current_battle.finished:
+        return {self.possible_agents[0]: {}, self.possible_agents[1]: {}}
+
+    def step(self, actions: Dict[str, ActionType]) -> Tuple[
+        Dict[str, ObsType],
+        Dict[str, float],
+        Dict[str, bool],
+        Dict[str, bool],
+        Dict[str, Dict[str, Any]],
+    ]:
+        assert self.current_battle1 is not None
+        assert self.current_battle2 is not None
+        if self.current_battle1.finished:
             raise RuntimeError("Battle is already finished, call reset")
-        battle = copy.copy(self.current_battle)
-        battle.logger = None
-        self.last_battle = battle
-        self._actions.put(action)
-        observation = self._observations.get()
-        reward = self.calc_reward(self.last_battle, self.current_battle)
+        battle1 = copy.copy(self.current_battle1)
+        battle1.logger = None
+        battle2 = copy.copy(self.current_battle2)
+        battle2.logger = None
+        self.last_battle1 = battle1
```

Comment:
Sorry for the random (and potentially ignorant) question - do we need to store `last_battle`s in the env? I ask for two reasons:
1. My understanding is `copy.copy` won't deep copy `Pokemon` objects, so `calc_reward` won't work properly because the Pokemon objects will be the same across battle copies (e.g. if a pokemon faints in this turn recorded by `battle1`, it will also appear fainted in `last_battle1`); I don't think this is tested either. Deep Copy is also suuuuuper slow, and so there could be a big tradeoff here
2. IIUC, the difference in reward isn't a necessary reward design; e.g. it makes sense to just reward agents for reaching an end state. In this case, it would make sense to allow the user to save `last_battle` in their own custom `Player` class and access it when they need to define the reward as a difference between two states

Im also not very confident in either #1 or #2, so sorry if this is random noise haha

Reply:
I actually agree entirely. My reasoning is that if someone wants the last battle, they can store it themselves, we shouldn't be going through the trouble. That's coming in the follow-up PR :) I just wanted to make this diff as minimal and trustable as possible.